packages/react/src/primitives/DropZone/useDropZone.ts

```diff
@@ -0,0 +1,116 @@
+import { useState } from 'react';
+import { UseDropZoneProps, UseDropZoneReturn } from './types';
+import { isFunction } from '@aws-amplify/ui';
+
+type DragFile =
+  | {
+      kind: string;
+      type: string;
+    }
+  | File;
+
+function filterAllowedFiles<FileType extends DragFile = DragFile>(
+  files: FileType[],
+  acceptedFileTypes: string[]
+): { accepted: FileType[]; rejected: FileType[] } {
+  // Allow any files if acceptedFileTypes is undefined, empty array, or contains '*'
+  if (
+    !acceptedFileTypes ||
+    acceptedFileTypes.length === 0 ||
+    acceptedFileTypes.includes('*')
+  ) {
+    return { accepted: files, rejected: [] };
+  }
+  const accepted: FileType[] = [];
+  const rejected: FileType[] = [];
+
+  function filterFile({ type = '' }) {
+    const mimeType = type.toLowerCase();
+    const baseMimeType = mimeType.split('/')[0];
+
+    return acceptedFileTypes.some((type) => {
+      const validType = type.trim().toLowerCase();
+      if (validType.endsWith('/*')) {
+        // This is something like a image/* mime type
+        return baseMimeType === validType.split('/')[0];
+      }
+      return mimeType === validType;
+    });
+  }
+
+  files.forEach((file) => {
+    (filterFile(file) ? accepted : rejected).push(file);
+  });
+
+  return { accepted, rejected };
+}
+
+export function useDropZone({
+  onDropComplete,
+  acceptedFileTypes = [],
+}: UseDropZoneProps): UseDropZoneReturn {
+  const [isDragActive, setIsDragActive] = useState(false);
+  const [isDragAccept, setIsDragAccept] = useState(false);
+  const [isDragReject, setIsDragReject] = useState(false);
```

Comment:
One question I have is should this just be a single enum state variable like `dragState` which has type `'default' | 'accept' | 'reject'` @hbuchel @calebpollman @reesscot

Reply:
Definitely think we should go with a string union since the booleans conflict with eachother